### PR TITLE
dynamically resize the table matrix on set_value

### PR DESCRIPTION
**Pull request by hiaselhans**
I will conform to pr checklist later.

here is a use case:
``` python
sheet=pyexcel.Sheet()
sheet[1,1] = "JO"
sheet["AA1"] = "test"
```

that leads to an index error.
I think it is a valid use case to set arbitrary cells at random positions?

This pr adds arguments to uniform so we can make an array of min_width and min_height

```

---------------------------------------------------------------------------
IndexError                                Traceback (most recent call last)
<ipython-input-2-df6d362a346f> in <module>
      1 sheet=pyexcel.Sheet()
      2 sheet[1,1] = "JO"
----> 3 sheet["AA1"] = "test"

~/dev/pyexcel/pyexcel/sheet.py in __setitem__(self, aset, c)
    607             self.cell_value(row, column, c)
    608         else:
--> 609             Matrix.__setitem__(self, aset, c)
    610 
    611     def __len__(self):

~/dev/pyexcel/pyexcel/internal/sheets/matrix.py in __setitem__(self, aset, cell_value)
    467         elif isinstance(aset, str):
    468             row, column = utils.excel_cell_position(aset)
--> 469             return self.cell_value(row, column, cell_value)
    470         else:
    471             raise IndexError

~/dev/pyexcel/pyexcel/internal/sheets/matrix.py in cell_value(self, row, column, new_value)
     94                 raise IndexError("Index out of range")
     95             else:
---> 96                 self.paste((row, column), [[new_value]])
     97 
     98     def row_at(self, index):

~/dev/pyexcel/pyexcel/internal/sheets/matrix.py in paste(self, topleft_corner, rows, columns)
    405         """
    406         if rows:
--> 407             self._paste_rows(topleft_corner, rows)
    408         elif columns:
    409             self._paste_columns(topleft_corner, columns)

~/dev/pyexcel/pyexcel/internal/sheets/matrix.py in _paste_rows(self, topleft_corner, rows)
    427             set_index = starting_row + index
    428             if set_index < number_of_rows:
--> 429                 self._set_row_at(set_index, row, starting=topleft_corner[1])
    430             else:
    431                 real_row = [constants.DEFAULT_NA] * topleft_corner[1] + row

~/dev/pyexcel/pyexcel/internal/sheets/matrix.py in _set_row_at(self, row_index, data_array, starting)
    153             self.__width, self.__array = uniform(self.__array)
    154         else:
--> 155             raise IndexError(constants.MESSAGE_INDEX_OUT_OF_RANGE)
    156 
    157     def _extend_row(self, row):
```

With your PR, here is a check list:

- [ ] Has test cases written?
- [ ] Has all code lines tested?
- [ ] Has `make format` been run?
- [ ] Please update CHANGELOG.yml(not CHANGELOG.rst)
- [ ] Passes all Travis CI builds
- [ ] Has fair amount of documentation if your change is complex
- [x] Agree on NEW BSD License for your contribution

### CODE REVIEW

**Comment by chfw** [comment]
Yep, pyexcel has not explored random access, which restrict use cases like yours. 

I recommend you have a closer look at paste function, which I explicitly forbid it exceeding the original dimension. You may find a way to relax it.

**Comment by hiaselhans** [comment]
i see that, the restriction is in _set_row_at.

so, given that there is an expensive call to uniform() in that function too, is there something wrong with my approach or is it just unintended behavior to allow random inserts?

**Comment by chfw** [comment]
I could be biased. It seems uniform function will also grow the two dimensional array, plus the original intent. Could the growth function be separated?

As to other side effect, I don’t think so. Current constraints were kept because personally I do not have a strong use case to support it. I wrote a lot of useless functions in pyexcel in the beginning.

So, as long as your change would pass the existing unit tests, your PR with unit tests for new code are good enough to be released.

**Comment by chfw** [comment]
‘make format’ will do automatically code formatting.

**Comment by chfw** [comment]
Good, it means there is no regression caused. 

Please write the unit tests which will protect your use case in the future:

1. test(s) that prove your use case works correct
2. exploratory tests if any

In the long run, the unit tests you contributed will live longer than the code you wrote. 

In essence, your use case will be kept alive as long as your unit tests will be kept. The non-test codes you contributed are mortal.

**Comment by hiaselhans** [comment]
thx @chfw !

I will prepare this pr right and add tests.
To my understanding we don't have any additional overhead introduced by this.

**Comment by chfw** [comment]
As a side topic, TDD would advise you to write unit test first then write the code.

**Comment by hiaselhans** [comment]
sure!

I am happy to comply with your rules (even 80char limits) and contribute more, if you are fine with human beings like me having different ways of solving their problems :)

Shall i add a howto contribute and missing commands from pr template to ease the path for future contributors?

**Comment by chfw** [comment]
You are welcome to contribute more. 

Regarding contributor guide and PR template, you can find them here: https://github.com/pyexcel/pyexcel-mobans/tree/master/templates. Once your changes is committed there, `moban` command will semi-automatically reflect your changes in pyexcel and pyexcel's other repos

**Comment by hiaselhans** [comment]
thanks @chfw .

i meant to make a guide on tooling / requirements for pre-commit checks...

however thats a different story and i think all issues with this pr are resolved.
codecov should be increased after all...

**Comment by chfw** [comment]
Precommit sounds cool. I never had the time to work it out.

File: tests/test_sheet_access.py [hunk 1]
```diff
@@ -0,0 +1,18 @@
+import random
+
+from base import SheetBaseTestcase
+
+
+class SheetAccessTest(SheetBaseTestcase):
```
Comment (chfw) [inline tests/test_sheet_access.py:6]: Seems you come from OOP background!

Comment (chfw) [reply tests/test_sheet_access.py:6]: Every time I make a class, I will ask myself: will I use any of the OOP features? If answer is NO, I will not use use classes.

Comment (chfw) [reply tests/test_sheet_access.py:6]: Please do not take it as criticism. Developers have their own view. I would like to hear more from you. That is an exchange.

Comment (hiaselhans) [reply tests/test_sheet_access.py]: well, i'm not coming from the java world if that's what you mean... ;)

If you have a hard opinion about it i can revert it to pure functions, but especially in unit tests i liked the idea of having a common base class. For grouping as well as for the common setUp and tearDown functions. I saw you were using them quite frequently too...

Comment (chfw) [reply tests/test_sheet_access.py]: Yes, I am opinionated when it comes to Java. “Thinking in Java” book had a a whole chapter: everything is object, hence Java developers cannot forget it when they write Python code.

Surely if you use OOP features, then a class will make more sense. What I am object to make a class for the sake of OOP.

Comment (hiaselhans) [reply tests/test_sheet_access.py]: sure.

i added another test to raise the coverage, and make the point of having a class more clear ;)
Don't know why codecov still fails. any idea?

**Comment by chfw** [comment]
https://codecov.io/gh/pyexcel/pyexcel/src/f26d77edf8807a07b205367589371e538bd7c497/tests/test_sheet_access.py
Screenshot 2020-11-05 at 12 23 06

**Comment by hiaselhans** [comment]
well. shame on me. the regex seems to be (Test*)

fixed that. now theres a ton of commits but you can squash them if you want..

File: pyexcel/internal/sheets/matrix.py [hunk 2]
```diff
@@ -82,18 +82,18 @@ def cell_value(self, row, column, new_value=None):
...
-                # set
-                self.__array[row][column] = new_value
-        else:
-            if new_value is None:
                 raise IndexError("Index out of range")
-            else:
-                self.paste((row, column), [[new_value]])
+        else:
+            if not fit:
+                width, array = uniform(self.__array, row+1, column+1)
```
Comment (chfw) [inline pyexcel/internal/sheets/matrix.py:93]: I think here, it is supposed that row > number of rows and column > number of rows.

what happens if: 

1. row < number of rows but column > number of rows
2. row > number of rows but column < number of rows

Comment (chfw) [reply pyexcel/internal/sheets/matrix.py:93]: it is good to add two tests to test above two conditions

**Comment by chfw** [comment]
the PR looks good now. Let's see about the 'extra miles' if you could extend the unit tests a bit.

File: pyexcel/internal/sheets/matrix.py (hunk 2 above)
Comment (hiaselhans) [reply pyexcel/internal/sheets/matrix.py:93]: i added those two edge cases as test, but to me it looks pretty clear:

```
fit = row < row_num and column < column_num
row < row_num, column > column_num => True and False => fit = false
row > row_num, column < coulmn_num => False and True => fit = false
```

**Comment by chfw** [comment]
https://github.com/pyexcel/pyexcel/commit/da3d6820424e6ad36c72e2a644caa7f8b098a2fb
